internal/docstore/firedocstore/fs.go

```diff
@@ -487,6 +506,38 @@ func (c *collection) updateWrites(doc driver.Document, docName string, mods []dr
 	return []*pb.Write{w}, nil
 }
 
+func (c *collection) runActionsUnordered(ctx context.Context, actions []*driver.Action) driver.ActionListError {
+	// Split into groups the same way, but run them concurrently.
+	// TODO(jba): group without considering order.
+	groups := driver.SplitActions(actions, shouldSplit)
+	errs := make([]error, len(actions))
+	var wg sync.WaitGroup
+	groupBaseIndex := 0 // index in actions of first action in group
+	for _, g := range groups {
+		g := g
+		base := groupBaseIndex
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			if g[0].Kind == driver.Get {
+				errs := c.runGets(ctx, g)
+				for i, err := range errs {
+					errs[base+i] = err
+				}
+			} else {
+				err := c.runWrites(ctx, g)
+				// Writes run in a transaction, so there is a single error for the group.
+				for i := 0; i < len(g); i++ {
+					errs[base+i] = err
```

Comment:
Why do we need to repeat the same errors and return them? It doesn't seem to help the user this way, instead, we could just return all different ones.

Reply:
Unordered lists have to return an error for each failed action. In this case, all the actions in the write failed. We have to a return an error for all of them.

How would we make the errors different? The only thing that's different is the action itself, but the index lets the user look up the action, so we are already encoding that information.